File: frontend/src/types/itemDemand.js
/**
 * Types and utilities for item demand summary.
 * Matches backend Pydantic schemas from API-101.
 */

/**
 * @typedef {Object} LinkedSalesOrder
 * @property {number} id
 * @property {string} code
 * @property {string|null} customer
 */

/**
 * @typedef {Object} AllocationDetail
 * @property {'production_order'} type
 * @property {string} reference_code
 * @property {number} reference_id
 * @property {number} quantity
 * @property {string|null} needed_date
 * @property {string} status
 * @property {LinkedSalesOrder|null} linked_sales_order
 */

/**
 * @typedef {Object} IncomingDetail
 * @property {'purchase_order'} type
 * @property {string} reference_code
 * @property {number} reference_id
 * @property {number} quantity
 * @property {string|null} expected_date
 * @property {string} status
 * @property {string|null} vendor
 */

/**
 * @typedef {Object} ShortageInfo
 * @property {boolean} is_short
 * @property {number} quantity
 * @property {string[]} blocking_orders
 */

/**
 * @typedef {Object} QuantitySummary
 * @property {number} on_hand
 * @property {number} allocated
 * @property {number} available
 * @property {number} incoming
 * @property {number} projected
 */

/**
 * @typedef {Object} ItemDemandSummary
 * @property {number} item_id
 * @property {string} sku
 * @property {string} name
 * @property {QuantitySummary} quantities
 * @property {AllocationDetail[]} allocations
 * @property {IncomingDetail[]} incoming
 * @property {ShortageInfo} shortage
 */

/**
 * Stock status types for visual indicators.
 * @typedef {'healthy' | 'tight' | 'short' | 'critical'} StockStatus
 */

/**
 * Determine stock status from quantities.
 * @param {QuantitySummary} quantities
 * @returns {StockStatus}
 */
export function getStockStatus(quantities) {
  const { available, on_hand } = quantities;

  // Parse as numbers in case they're strings from JSON
  const availNum = parseFloat(available);
  const onHandNum = parseFloat(on_hand);

  if (availNum < 0) {
    return 'critical';  // Negative available = shortage
  }
  if (availNum === 0) {
    return 'short';     // Zero available
  }
  if (onHandNum > 0 && availNum < onHandNum * 0.2) {
    return 'tight';     // Less than 20% available
  }
  return 'healthy';
}

/**
 * Get Tailwind color classes for stock status.
 * @param {StockStatus} status
 * @returns {{bg: string, text: string, border: string, dot: string}}
 */
export function getStatusColors(status) {
  switch (status) {
    case 'critical':
      return {
        bg: 'bg-red-900/30',
        text: 'text-red-400',
        border: 'border-red-700',
        dot: 'bg-red-500'
      };
    case 'short':
      return {
        bg: 'bg-orange-900/30',
        text: 'text-orange-400',
        border: 'border-orange-700',
        dot: 'bg-orange-500'
      };
    case 'tight':
      return {
        bg: 'bg-yellow-900/30',
        text: 'text-yellow-400',
        border: 'border-yellow-700',
        dot: 'bg-yellow-500'
      };
    case 'healthy':
    default:
      return {
        bg: 'bg-green-900/30',
        text: 'text-green-400',
        border: 'border-green-700',
        dot: 'bg-green-500'
      };
  }
}

File: backend/app/services/item_demand.py
"""
Service layer for item demand calculations.

Provides functions to calculate:
- On-hand quantity from inventory
- Allocations from active production orders
- Incoming supply from open purchase orders
- Shortage detection
"""
from datetime import date
from decimal import Decimal
from typing import Optional, List
from sqlalchemy import func
from sqlalchemy.orm import Session

from app.models.product import Product
from app.models.bom import BOM, BOMLine
from app.models.production_order import ProductionOrder
from app.models.sales_order import SalesOrder
from app.models.purchase_order import PurchaseOrder, PurchaseOrderLine
from app.models.inventory import Inventory
from app.schemas.item_demand import (
    ItemDemandSummary, QuantitySummary, AllocationDetail,
    IncomingDetail, ShortageInfo, LinkedSalesOrder
)


def get_item_on_hand(db: Session, product_id: int) -> Decimal:
    """
    Get total on-hand quantity for a product across all locations.

    Uses the Inventory table which tracks current stock levels.
    """
    result = db.query(
        func.coalesce(func.sum(Inventory.on_hand_quantity), 0)
    ).filter(
        Inventory.product_id == product_id
    ).scalar()

    return Decimal(str(result or 0))


def get_allocated_quantity(db: Session, product_id: int) -> Decimal:
    """
    Get total allocated quantity for a product from active production orders.

    Calculates allocation based on active work orders that consume this
    product as a BOM component. This is the accurate allocation vs the
    Inventory.allocated_quantity column which may be stale.
    """
    # Find all BOMs that use this product as a component
    bom_lines_using_product = db.query(
        BOMLine.bom_id,
        BOMLine.quantity
    ).filter(
        BOMLine.component_id == product_id
    ).subquery()

    # Get the parent products from those BOMs
    boms_using_product = db.query(
        BOM.product_id,
        bom_lines_using_product.c.quantity
    ).join(
        bom_lines_using_product,
        BOM.id == bom_lines_using_product.c.bom_id
    ).filter(
        BOM.active == True  # noqa: E712
    ).subquery()

    # Active production order statuses
    active_statuses = ['draft', 'released', 'scheduled', 'in_progress']

    # Sum up allocation from all active production orders
    result = db.query(
        func.coalesce(
            func.sum(
                (ProductionOrder.quantity_ordered -
                 func.coalesce(ProductionOrder.quantity_completed, 0) -
                 func.coalesce(ProductionOrder.quantity_scrapped, 0)) *
                boms_using_product.c.quantity
            ),
            0
        )
    ).select_from(ProductionOrder).join(
        boms_using_product,
        ProductionOrder.product_id == boms_using_product.c.product_id
    ).filter(
        ProductionOrder.status.in_(active_statuses)
    ).scalar()

    return Decimal(str(result or 0))


def get_production_order_allocations(db: Session, product_id: int) -> List[AllocationDetail]:
    """
    Get all active production orders that consume this product.

    Finds products that use this item in their BOM, then finds
    production orders for those products that are in active status.

    Active statuses: draft, released, scheduled, in_progress
    """
    # Find all BOMs that use this product as a component
    bom_lines_using_product = db.query(
        BOMLine.bom_id,
        BOMLine.quantity
    ).filter(
        BOMLine.component_id == product_id
    ).subquery()

    # Get the parent products from those BOMs
    boms_using_product = db.query(
        BOM.product_id,
        bom_lines_using_product.c.quantity
    ).join(
        bom_lines_using_product,
        BOM.id == bom_lines_using_product.c.bom_id
    ).filter(
        BOM.active == True  # noqa: E712
    ).subquery()

    # Active production order statuses (not complete, closed, or cancelled)
    active_statuses = ['draft', 'released', 'scheduled', 'in_progress']

    # Find production orders for products that use this component
    query = db.query(
        ProductionOrder,
        boms_using_product.c.quantity.label('qty_per_unit')
    ).join(
        boms_using_product,
        ProductionOrder.product_id == boms_using_product.c.product_id
    ).filter(
        ProductionOrder.status.in_(active_statuses)
    ).order_by(
        ProductionOrder.due_date.asc().nullslast()
    )

    results = query.all()

    allocations = []
    for production_order, qty_per_unit in results:
        # Calculate quantity needed: (ordered - completed - scrapped) * qty_per_unit
        remaining = (
            (production_order.quantity_ordered or Decimal("0")) -
            (production_order.quantity_completed or Decimal("0")) -
            (production_order.quantity_scrapped or Decimal("0"))
        )
        allocated_qty = remaining * Decimal(str(qty_per_unit))

        if allocated_qty <= 0:
            continue  # Skip if no allocation needed

        # Get linked sales order if any
        linked_so = None
        if production_order.sales_order_id:
            so = db.query(SalesOrder).filter(
                SalesOrder.id == production_order.sales_order_id
            ).first()
            if so:
                linked_so = LinkedSalesOrder(
                    id=so.id,
                    code=so.order_number,
                    customer=so.customer_name or "Unknown"
                )

        allocations.append(AllocationDetail(
            type="production_order",
            reference_code=production_order.code,
            reference_id=production_order.id,
            quantity=allocated_qty,
            needed_date=production_order.due_date,
            status=production_order.status,
            linked_sales_order=linked_so
        ))

    return allocations


def get_incoming_supply(db: Session, product_id: int) -> List[IncomingDetail]:
    """
    Get all open purchase order lines for this product.

    Open PO statuses: draft, ordered, shipped (not received, closed, cancelled)
    """
    # Active purchase order statuses
    active_statuses = ['draft', 'ordered', 'shipped']

    # Query purchase order lines for this product
    query = db.query(
        PurchaseOrderLine,
        PurchaseOrder,
    ).join(
        PurchaseOrder,
        PurchaseOrderLine.purchase_order_id == PurchaseOrder.id
    ).filter(
        PurchaseOrderLine.product_id == product_id,
        PurchaseOrder.status.in_(active_statuses)
    ).order_by(
        PurchaseOrder.expected_date.asc().nullslast()
    )

    results = query.all()

    incoming = []
    for pol, po in results:
        # Calculate remaining to receive
        ordered = pol.quantity_ordered or Decimal("0")
        received = pol.quantity_received or Decimal("0")
        remaining = ordered - received

        if remaining <= 0:
            continue  # Fully received

        # Get vendor name
        vendor_name = None
        if po.vendor:
            vendor_name = po.vendor.name

        incoming.append(IncomingDetail(
            type="purchase_order",
            reference_code=po.po_number,
            reference_id=po.id,
            quantity=remaining,
            expected_date=po.expected_date,
            status=po.status,
            vendor=vendor_name
        ))

    return incoming


def calculate_shortage(
    available: Decimal,
    allocations: List[AllocationDetail]
) -> ShortageInfo:
    """
    Determine if there's a shortage and which orders are affected.

    A shortage exists when available < 0 (allocated > on_hand).
    """
    if available >= 0:
        return ShortageInfo(
            is_short=False,
            quantity=Decimal("0"),
            blocking_orders=[]
        )

    # There's a shortage
    shortage_amount = abs(available)

    # Find which orders are blocked (simplified: all orders if there's a shortage)
    # More sophisticated: could calculate which specific orders can't be fulfilled
    # based on priority/date order
    blocking = []

    # Sort allocations by priority (date, then status)
    sorted_allocs = sorted(
        allocations,
        key=lambda a: (a.needed_date or date.max, a.status != 'released')
    )

    # Walk through allocations to determine which are blocked
    running_total = Decimal("0")
    on_hand = available + sum(a.quantity for a in allocations)  # Calculate original on_hand

    for alloc in sorted_allocs:
        running_total += alloc.quantity
        if running_total > on_hand:
            blocking.append(alloc.reference_code)

    return ShortageInfo(
        is_short=True,
        quantity=shortage_amount,
        blocking_orders=blocking
    )


def get_item_demand_summary(db: Session, item_id: int) -> Optional[ItemDemandSummary]:
    """
    Get complete demand summary for a product.

    Returns None if product doesn't exist.

    The summary includes:
    - Current on-hand quantity
    - Allocated quantity (from active production orders)
    - Available quantity (on_hand - allocated)
    - Incoming quantity (from open purchase orders)
    - Projected quantity (available + incoming)
    - List of allocations with linked sales orders
    - List of incoming supply
    - Shortage information if applicable
    """
    # Get the product
    product = db.query(Product).filter(Product.id == item_id).first()
    if not product:
        return None

    # Calculate quantities
    on_hand = get_item_on_hand(db, item_id)
    allocations = get_production_order_allocations(db, item_id)
    incoming_list = get_incoming_supply(db, item_id)

    allocated = sum(a.quantity for a in allocations)
    incoming = sum(i.quantity for i in incoming_list)
    available = on_hand - allocated
    projected = available + incoming

    shortage = calculate_shortage(available, allocations)

    return ItemDemandSummary(
        item_id=product.id,
        sku=product.sku,
        name=product.name,
        stocking_policy=product.stocking_policy or "on_demand",
        reorder_point=product.reorder_point,
        quantities=QuantitySummary(
            on_hand=on_hand,
            allocated=allocated,
            available=available,
            incoming=incoming,
            projected=projected
        ),
        allocations=allocations,
        incoming=incoming_list,
        shortage=shortage
    )
